### Chapter 3 - DS for Public Policy
- Exploration of solar data from the National Institute of Standards and Technology (NIST) Net Zero Energy Residential Test Facility, a house that produces as much energy as it uses in a year.
- Looking at a slice of the net zero house's photovoltaic time series data from 2015. The dataset contains hourly estimates of solar energy production and exposure on the Net Zero home's solar panels.
- We want to know how much energy is produced a year and how variable that production is over the course of a year

In [1]:
# Import statements of necessary libraries
import polars as plr
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
import os

In [ ]:
# Change the working directory to our desired path
os.chdir("Public Policy/")
# Load the solar data from a CSV file into a Polars DataFrame
solar = plr.read_csv("data/PV-hour.csv")
# Print the shape of the DataFrame
print(solar.shape)
# View first 10 rows of the DataFrame
solar.head(10)

(8760, 32)


shape: (10, 32)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ Timestamp ┆ PV_PVSyst ┆ PV_PVSyst ┆ PV_PVBack ┆ … ┆ PV_String ┆ PV_PVSyst ┆ PV_PVSyst ┆ PV_PVIns │
│ ---       ┆ em1ACEner ┆ em2ACEner ┆ sideTemp2 ┆   ┆ CurrentIS ┆ em1ACPowe ┆ em2ACPowe ┆ olationH │
│ str       ┆ gyOSEACPV ┆ gyOSEACPV ┆ ---       ┆   ┆ tr4       ┆ rOSPACPV1 ┆ rOSPACPV2 ┆ Array    │
│           ┆ 1OS       ┆ 2OS       ┆ f64       ┆   ┆ ---       ┆ OS        ┆ OS        ┆ ---      │
│           ┆ ---       ┆ ---       ┆           ┆   ┆ f64       ┆ ---       ┆ ---       ┆ f64      │
│           ┆ f64       ┆ f64       ┆           ┆   ┆           ┆ f64       ┆ f64       ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -10.14399 ┆ … ┆ 0.000032  ┆ 1.247293  ┆ 1.524335  ┆ 0.057408 │
│ 1 00:00:0 ┆           ┆           ┆ 4         ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -10.10454 ┆ … ┆ 0.000033  ┆ 1.246217  ┆ 1.595343  ┆ 0.060286 │
│ 1 01:00:0 ┆           ┆           ┆ 5         ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -9.213759 ┆ … ┆ 0.000027  ┆ 1.240928  ┆ 1.616503  ┆ 0.058811 │
│ 1 02:00:0 ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -8.030592 ┆ … ┆ 0.000025  ┆ 1.254152  ┆ 1.629727  ┆ 0.058865 │
│ 1 03:00:0 ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -5.850926 ┆ … ┆ 0.00003   ┆ 1.256797  ┆ 1.611213  ┆ 0.059288 │
│ 1 04:00:0 ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -5.063146 ┆ … ┆ 0.000029  ┆ 1.251507  ┆ 1.650886  ┆ 0.058733 │
│ 1 05:00:0 ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 0.0       ┆ 0.0       ┆ -4.698231 ┆ … ┆ 0.000148  ┆ 1.264732  ┆ 1.637662  ┆ 0.059928 │
│ 1 06:00:0 ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 12.0      ┆ 11.0      ┆ -3.238496 ┆ … ┆ 0.034785  ┆ 12.381218 ┆ 14.256448 ┆ 0.599965 │
│ 1 07:00:0 ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 511.0     ┆ 504.0     ┆ 0.685583  ┆ … ┆ 0.629558  ┆ 501.58333 ┆ 500.28998 ┆ 5.740961 │
│ 1 08:00:0 ┆           ┆           ┆           ┆   ┆           ┆ 4         ┆ 1         ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2015-02-0 ┆ 1192.0    ┆ 1179.0    ┆ 5.329179  ┆ … ┆ 0.840221  ┆ 683.25047 ┆ 677.55602 ┆ 7.538642 │
│ 1 09:00:0 ┆           ┆           ┆           ┆   ┆           ┆ 8         ┆ 2         ┆          │
│ 0-05:00   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

We are interested in the total amount of sunlight shining on the solar arrays at any given hour (kWh), and will focus on the `PV_PVInsolationHArray` in the DataFrame `solar`. 

In [3]:
# View the 5 number summary of the PV_PVInsolationHArray
solar[["PV_PVInsolationHArray"]].describe()

shape: (9, 2)
┌────────────┬───────────────────────┐
│ statistic  ┆ PV_PVInsolationHArray │
│ ---        ┆ ---                   │
│ str        ┆ f64                   │
╞════════════╪═══════════════════════╡
│ count      ┆ 8760.0                │
│ null_count ┆ 0.0                   │
│ mean       ┆ 9.452628              │
│ std        ┆ 14.533754             │
│ min        ┆ 0.0                   │
│ 25%        ┆ 0.055849              │
│ 50%        ┆ 0.339645              │
│ 75%        ┆ 14.842793             │
│ max        ┆ 54.843525             │
└────────────┴───────────────────────┘

So, we can view the hourly variability in this basic summary. This tells us that the PV arrays are exposed to fairly small amounts of energy for most hours as indicated by the median relative to the mean. However, there are periods where the PV arrays are exposed to intense sunlight.

We can measure the variability by using the coefficient of variation (CV) by dividing the standard deviation, $\sigma$, in relation to the mean, $\mu$:

$CV = \frac{\sigma}{\mu}$

In [4]:
# Calculate the coefficient of variation (CV) for the PV_PVInsolationHArray
cv = solar["PV_PVInsolationHArray"].std() / solar["PV_PVInsolationHArray"].mean()
print(f"Coefficient of Variation (CV): {cv}")

Coefficient of Variation (CV): 1.5375357839988257


Values of the coefficient of variation that exceed 1 indicate greater dispersion, and this indicates one standard deviation is 1.54 times (rounded) as wide as the mean - variable energy generation.

Can we view this with a line graph?

In [5]:
# Convert the 'Timestamp' column to datetime format
solar.with_columns(
    plr.col("Timestamp").str.to_datetime("%Y-%m-%d %H:%M:%S", strict=False)
)
# Create a line plot using Plotly
fig = px.line(solar.to_pandas(), x="Timestamp", y="PV_PVInsolationHArray",
              title="Hourly PV Insolation on Solar Arrays",
              labels={"PV_PVInsolationHArray": "PV Insolation (kWh)", "Timestamp": "Time"})
fig.show()

High variability... will come back to this later, but for now we will leave this work behind.